[GTK][WPE] Network sessions are not destroyed in platformFinalize

pgorszkowski-igalia opened the pull request:
#### 34015b17e69f46303cae04d174933c86544ff3e5
<pre>
[GTK][WPE] Network sessions are not destroyed in platformFinalize
<a href="https://bugs.webkit.org/show_bug.cgi?id=281143">https://bugs.webkit.org/show_bug.cgi?id=281143</a>

Reviewed by Michael Catanzaro and Carlos Garcia Campos.

When we close UIProcess it sends to NetworkProcess message(s)
to destroy all sessions (DestroySession) but right after that
the IPC connection is close which causes that NetworkProcess
can be closed without destroying all sessions.

This change destroys all sessions when the NetworkProcess is
closing.

Original author: Andrzej Surdej &lt;Andrzej_Surdej@comcast.com&gt;

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1406">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1406</a>

* Source/WebKit/NetworkProcess/soup/NetworkProcessMainSoup.cpp:

Canonical link: <a href="https://commits.webkit.org/285119@main">https://commits.webkit.org/285119@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0503f2c1c868b02d7634c74290a42d95f6bc069

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71238 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50651 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24011 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75346 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22443 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58451 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22262 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56310 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14769 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74304 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46013 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61390 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36745 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42686 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20784 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64577 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19215 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77068 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15472 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18401 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64283 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15514 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61425 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64007 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15821 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12149 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5780 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46451 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1230 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47522 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47264 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/NetworkProcess/curl/NetworkProcessMainCurl.cpp:
```diff
@@ -35,6 +35,12 @@ class NetworkProcessMainCurl final: public AuxiliaryProcessMainBaseNoSingleton<N
+            sessionIDs.append(session.sessionID());
+        });
+        for (auto& sessionID : sessionIDs)
+            process().destroySession(sessionID);
```
GTK and WPE ports don't use curls, so either split the patch into curl/soup or change the bug title.

pgorszkowski-igalia replied:
ok, will create a separate PR and bug report for CURL case

carlosgcampos commented on Source/WebKit/NetworkProcess/soup/NetworkProcessMainSoup.cpp:
```diff
@@ -49,9 +49,12 @@ class NetworkProcessMainSoup final: public AuxiliaryProcessMainBaseNoSingleton<N
+        });
+        for (auto& sessionID : sessionIDs)
+            process().destroySession(sessionID);
         process().destroySession(PAL::SessionID::defaultSessionID());
```
I guess the default session has already been destroyed at this point.

pgorszkowski-igalia replied:
It should be, you are right

mcatanzaro reviewed (approved):
Looks good to me now.

After all, what could go wrong. :)

(Personally I would have fixed soup and curl in this one pull request and just retitled it, but whatever. Two pull requests is fine too.)

carlosgcampos commented on Source/WebKit/NetworkProcess/soup/NetworkProcessMainSoup.cpp:
```diff
@@ -49,10 +49,12 @@ class NetworkProcessMainSoup final: public AuxiliaryProcessMainBaseNoSingleton<N
     {
-        // FIXME: Is this still needed? We should probably destroy all existing sessions at this point instead.
-        // Needed to destroy the SoupSession and SoupCookieJar, e.g. to avoid
-        // leaking SQLite temporary journaling files.
```
The fixme is no longer needed, but this comment is still valid.

pgorszkowski-igalia committed:
[GTK][WPE] Network sessions are not destroyed in platformFinalize
https://bugs.webkit.org/show_bug.cgi?id=281143

Reviewed by Michael Catanzaro and Carlos Garcia Campos.

When we close UIProcess it sends to NetworkProcess message(s)
to destroy all sessions (DestroySession) but right after that
the IPC connection is close which causes that NetworkProcess
can be closed without destroying all sessions.

This change destroys all sessions when the NetworkProcess is
closing.

Original author: Andrzej Surdej <Andrzej_Surdej@comcast.com>

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1406

* Source/WebKit/NetworkProcess/soup/NetworkProcessMainSoup.cpp:

Canonical link: https://commits.webkit.org/285119@main
